helix-term/src/ui/editor.rs

```diff
@@ -725,29 +571,36 @@ impl EditorView {
 
         let gutter_style = theme.get("ui.gutter");
         let gutter_selected_style = theme.get("ui.gutter.selected");
-
-        // avoid lots of small allocations by reusing a text buffer for each line
-        let mut text = String::with_capacity(8);
+        let gutter_style_virtual = theme.get("ui.gutter.virtual");
+        let gutter_selected_style_virtual = theme.get("ui.gutter.selected.virtual");
 
         for gutter_type in view.gutters() {
             let mut gutter = gutter_type.style(editor, doc, view, theme, is_focused);
             let width = gutter_type.width(view, doc);
-            text.reserve(width); // ensure there's enough space for the gutter
-            for (i, line) in (view.offset.row..(last_line + 1)).enumerate() {
-                let selected = cursors.contains(&line);
+            // avoid lots of small allocations by reusing a text buffer for each line
+            let mut text = String::with_capacity(width);
+            let cursors = cursors.clone();
+            let gutter_decoration = move |renderer: &mut TextRenderer, pos: LinePos| {
+                // TODO handle softwrap in gutters
+                let selected = cursors.contains(&pos.doc_line);
                 let x = viewport.x + offset;
-                let y = viewport.y + i as u16;
+                let y = viewport.y + pos.visual_line;
 
-                let gutter_style = if selected {
-                    gutter_selected_style
-                } else {
-                    gutter_style
+                let gutter_style = match (selected, pos.first_visual_line) {
+                    (false, false) => gutter_style,
+                    (true, false) => gutter_selected_style,
+                    (false, true) => gutter_style_virtual,
+                    (true, true) => gutter_selected_style_virtual,
```

Comment:
Noticed this while trying to theme my gutter: the matching on the second bool in the pair is flipped.  It should be:

```
(false, true) => gutter_style,
(true, true) => gutter_selected_style,
(false, false) => gutter_style_virtual,
(true, false) => gutter_selected_style_virtual,
```

As a result, right now when you have e.g. `ui.gutter.virtual` defined in your theme, the virtual/non-virtual gutter rows are flipped from what they should be: non-virtual gutter rows get the virtual color, and the virtual rows get the non-virtual color.

Reply:
Yup, no worries!  Happens to us all. :-)